src/transformers/pipelines/image_segmentation.py

```diff
@@ -104,39 +119,67 @@ def _forward(self, model_inputs):
-        keys = ["score", "label", "mask"]
-        annotation = [
-            dict(zip(keys, vals))
-            for vals in zip(raw_annotation["scores"], raw_annotation["labels"], raw_annotation["masks"])
-        ]
+    def postprocess(self, model_outputs, raw_image=False, threshold=0.9, mask_threshold=0.5):
+        if hasattr(self.feature_extractor, "post_process_segmentation"):
+            # Panoptic
+            raw_annotations = self.feature_extractor.post_process_segmentation(
+                model_outputs, model_outputs["target_size"], threshold=threshold, mask_threshold=0.5
+            )
+            raw_annotation = raw_annotations[0]
+            raw_annotation["masks"] *= 255  # [0,1] -> [0,255] black and white pixels
+            raw_annotation["scores"] = raw_annotation["scores"].tolist()
+            raw_annotation["labels"] = [self.model.config.id2label[label.item()] for label in raw_annotation["labels"]]
+            raw_annotation["masks"] = [
+                Image.fromarray(mask.numpy().astype(np.int8), mode="L") for mask in raw_annotation["masks"]
+            ]
+            if not raw_image:
+                raw_annotation["masks"] = [self._get_mask_str(mask) for mask in raw_annotation["masks"]]
+
+            # {"scores": [...], ...} --> [{"score":x, ...}, ...]
+            keys = ["score", "label", "mask"]
+            annotation = [
+                dict(zip(keys, vals))
+                for vals in zip(raw_annotation["scores"], raw_annotation["labels"], raw_annotation["masks"])
+            ]
+        else:
+            # Default logits
+            logits = model_outputs.logits
+            if len(logits.shape) != 4:
+                raise ValueError(f"Logits don't have expected dimensions, expected [1, N, H, W], got {logits.shape}")
+            # Softmax
+            logits = logits.log_softmax(dim=1)
+            batch_size, num_labels, height, width = logits.shape
+            expected_num_labels = len(self.model.config.id2label)
+            if num_labels != expected_num_labels:
+                raise ValueError(
+                    f"Logits don't have expected dimensions, expected [1, {num_labels}, H, W], got {logits.shape}"
+                )
+            size = model_outputs["target_size"].tolist()[0]
+            logits_reshaped = torchvision.transforms.Resize(size)(logits)
+            classes = logits_reshaped.argmax(dim=1)[0]
+            annotation = []
+            for label_id in range(num_labels):
+                label = self.model.config.id2label[label_id]
+                mask = classes == label_id
+                score = ((mask * logits_reshaped[0, label_id]).sum() / mask.sum()).exp().item()
+                mask = Image.fromarray(mask.numpy().astype(np.int8), mode="L")
+                if not raw_image:
+                    mask = self._get_mask_str(mask)
+                annotation.append({"score": score, "label": label, "mask": mask})
```

Comment:
This is where grouping the tasks under the same umbrella is dangerous. The output of the semantic segmentation should be one mask, not one per label.